transaction/core/src/membership_proofs/mod.rs

```diff
@@ -50,6 +50,139 @@ fn hash_nil() -> [u8; 32] {
+    // subtle::conditional_assign is assumed to be side-channel resistant
+    let mut result = TxOutMembershipElement {
+        range: Range {
+            from: a.range.from,
+            to: b.range.to,
+        },
+        hash: Default::default(),
+    };
+    result
+        .range
+        .from
+        .conditional_assign(&b.range.from, b_is_left);
+    result.range.to.conditional_assign(&a.range.to, b_is_left);
+
+    // Initialize left and right as if a is the left and b is the right,
+    // then use conditional_assign to fix it if not.
+    let mut left = a.hash.0;
+    let mut right = b.hash.0;
+    conditional_assign_32_bytes(&mut left, &b.hash.0, b_is_left);
+    conditional_assign_32_bytes(&mut right, &a.hash.0, b_is_left);
+
+    result.hash.0 = hash_nodes(&left, &right);
+    Ok(result)
+}
+
+// Helper: Conditionally assign to [u8; 32]
+//
+// It seems subtle doesn't implement ConditionallySelectable on [u8; 32], not sure why,
+// so we can't just use the conditional_assign function from that API.
+//
+// If this is slow then we can later used a faster implementation, but it probably won't be that slow.
+fn conditional_assign_32_bytes(target: &mut [u8; 32], src: &[u8; 32], cond: Choice) {
+    for idx in 0..32 {
+        target[idx].conditional_assign(&src[idx], cond);
+    }
+}
+
+/// Checks that a proof of membership is well-formed, and returns the hash that it implies for the root.
+/// This hash should then be checked against known root hash value.
+///
+/// Errors if:
+/// - Any proof elements have invalid ranges (from and to out of order)
+/// - The first element is missing or its range doesn't match proof.index
+/// - Any of the proof elements could not be combined with result of combining predecessors
+///
+/// Does not read the proof.highest_index value or compare it with the range
+/// that we compute by combining elements. But this should not be strictly necessary
+/// for secure merkle proof validation.
+///
+/// Note: This is pub in order to allow that it can be used in debugging assertions elsewhere.
+/// This could simply be a member function on TxOutMembershipProof, but that would require
+/// hash_nodes function to be in scope in that module, so for now we didn't do that.
+pub fn compute_implied_merkle_root(proof: &TxOutMembershipProof) -> Result<[u8; 32], Error> {
+    // All Ranges contained in the proof must be valid. An invalid Range could be created
+    // by deserializing invalid bytes.
+    if proof.elements.iter().any(|e| e.range.from > e.range.to) {
+        return Err(Error::RangeError(RangeError {}));
+    }
+
+    // The first element should correspond to proof.index in a well-formed proof
+    let first = proof
+        .elements
+        .first()
+        .ok_or_else(|| Error::MissingLeafHash(proof.index))?
+        .clone();
+    if first.range.from != proof.index || first.range.to != proof.index {
+        return Err(Error::MissingLeafHash(proof.index));
+    }
```

Comment:
(these checks would also go away if we just add this to `proof.first_hash` or something and make it not part of `proof.elements`